File: app/src/main/java/com/ocram/qichwadic/presentation/ui/activity/SearchActivity.java
package com.ocram.qichwadic.presentation.ui.activity;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.RelativeLayout;
import android.widget.SearchView;
import android.widget.Spinner;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.core.text.HtmlCompat;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.snackbar.Snackbar;
import com.ocram.qichwadic.QichwaDicApplication;
import com.ocram.qichwadic.R;
import com.ocram.qichwadic.domain.model.Definition;
import com.ocram.qichwadic.domain.model.SearchResult;
import com.ocram.qichwadic.presentation.ui.adapters.DictLangAdapter;
import com.ocram.qichwadic.presentation.ui.adapters.ResultListAdapter;
import com.ocram.qichwadic.presentation.ui.adapters.SpinnerResultAdapter;
import com.ocram.qichwadic.presentation.ui.custom.EndlessRecyclerViewScrollListener;
import com.ocram.qichwadic.presentation.model.DictLang;
import com.ocram.qichwadic.framework.preferences.PreferencesHelper;
import com.ocram.qichwadic.presentation.viewmodel.SearchViewModel;
import com.ocram.qichwadic.presentation.viewmodel.viewstate.SearchParams;
import com.ocram.qichwadic.presentation.viewmodel.viewstate.SearchViewState;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.OnClick;
import butterknife.OnItemSelected;
import io.reactivex.disposables.CompositeDisposable;

public class SearchActivity extends BaseActivity implements ResultListAdapter.DefinitionClickListener {

    @BindView(R.id.clSearch) CoordinatorLayout clSearch;
    @BindView(R.id.svSearch) SearchView svSearch;
    @BindView(R.id.spResults) Spinner spResults;
    @BindView(R.id.rvResults) RecyclerView rvResults;
    @BindView(R.id.searchProgressBar) ProgressBar searchProgressBar;
    @BindView(R.id.fetchMoreProgressBar) ProgressBar fetchMoreProgressBar;
    @BindView(R.id.llNoResults) LinearLayout llNoResults;
    @BindView(R.id.llSearchOptions) LinearLayout llSearchOptions;
    @BindView(R.id.spSearchTypes) Spinner spSearchTypes;
    @BindView(R.id.llResultsArea) RelativeLayout llResultsArea;
    @BindView(R.id.tvResultsTotal) TextView tvResultsTotal;
    @BindView(R.id.fabSearch) FloatingActionButton fabSearch;
    @BindView(R.id.ivSwapLanguages) ImageView ivSwapLanguages;
    @BindView(R.id.tvQichwaLang) TextView tvQichwaLang;
    @BindView(R.id.spDictLangs) Spinner spDictLangs;
    @BindView(R.id.vSeparator) View vSeparator;
    @BindView(R.id.btnNoResultsGetDictionaries) Button btnNoResultsGetDictionaries;
    private SpinnerResultAdapter spinnerResultAdapter;
    private ResultListAdapter resultListAdapter;
    private EndlessRecyclerViewScrollListener scrollListener;
    private int searchType;
    private Map<String, String> placeholders;
    private String quechuaPlaceholder;
    private boolean isFromQuechua;
    private SearchParams searchParams;
    private CompositeDisposable compositeDisposable;

    private SearchViewModel viewModel;

    @Inject
    PreferencesHelper preferencesHelper;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        preferencesHelper = QichwaDicApplication.get(this).getAppComponent().getPreferencesHelper();
        super.onCreate(savedInstanceState);
        this.compositeDisposable = new CompositeDisposable();
        loadPlaceholders();
        setAdapters();
        setSearchView();
        viewModel = ViewModelProviders.of(this).get(SearchViewModel.class);
        viewModel.getResultLiveData().observe(this, this::onResultsChanged);
        viewModel.getSearchViewState().observe(this, this::onSearchViewStateChanged);
        viewModel.getExtraDefinitions().observe(this, this::onFetchExtraDefinitions);
        viewModel.getLoadFetchMore().observe(this, this::onFetchMoreLoadingChanged);
        viewModel.getSaveFavoriteResult().observe(this, this::onSaveWordResult);
        this.initSavedSearchPrefs();
    }

    private void initSavedSearchPrefs() {
        if(this.searchParams == null) {
            compositeDisposable.add(
                    preferencesHelper.getLastSearchParams()
                            .doOnNext(this::setSearchParams)
                            .doOnTerminate(() -> svSearch.setQuery(svSearch.getQueryHint(), true))
                            .subscribe()
            );
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        scrollListener = null;
        if(compositeDisposable != null) {
            compositeDisposable.dispose();
        }
    }

    @Override
    protected int getLayoutId() {
        return R.layout.activity_main;
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_toolbar, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();

        if(id == R.id.nav_dictionaries) {
            openDictionariesGallery();
            return true;
        } else if(id == R.id.nav_favorites){
            Intent intent = new Intent(SearchActivity.this, FavoriteActivity.class);
            startActivity(intent);
            return true;
        }else if(id == R.id.nav_about){
            Intent intent = new Intent(SearchActivity.this, AboutActivity.class);
            startActivity(intent);
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void initViews() {
        setToolbar();
        setRecyclerViews();
    }

    private void setToolbar(){
        if(getSupportActionBar() != null){
            getSupportActionBar().setDisplayShowTitleEnabled(true);
            getSupportActionBar().setHomeButtonEnabled(false);
            getSupportActionBar().setDisplayHomeAsUpEnabled(false);
        }
    }

    private void setRecyclerViews(){
        LinearLayoutManager layoutManager = new LinearLayoutManager(this, RecyclerView.VERTICAL, false);
        rvResults.setLayoutManager(layoutManager);
        scrollListener = new EndlessRecyclerViewScrollListener(layoutManager) {
            @Override
            public void onLoadMore(int page, int totalItemsCount, RecyclerView view) {
                SearchResult currentSearchResult = (SearchResult)spResults.getSelectedItem();
                int totalToFetch = currentSearchResult.getTotal();
                if(totalToFetch > 20 && resultListAdapter.getItemCount() < totalToFetch){
                    int dictionaryId = currentSearchResult.getDictionaryId();
                    if(view != null){
                        viewModel.fetchMoreResults(dictionaryId, searchType, svSearch.getQuery().toString(), page + 1);
                    }
                }
            }
        };
        rvResults.addOnScrollListener(scrollListener);
    }

    @Override
    public void onItemShareClick(Definition definition) {
        String text = getString(R.string.share_definition_from_dictionary,
                definition.getDictionaryName(), definition.getWord(), definition.getMeaning());
        openShareIntent(HtmlCompat.fromHtml(text, HtmlCompat.FROM_HTML_MODE_LEGACY).toString());
    }

    @Override
    public void onItemFavoriteClick(Definition definition) {
        viewModel.saveFavorite(definition);
    }

    @OnClick(R.id.fabSearch)
    public void openFavorites(View v){
        if(svSearch.getQuery().length() > 0){
            svSearch.setQuery(svSearch.getQuery(), true);
        }
    }

    @OnClick(R.id.btnNoResultsGetDictionaries)
    public void openDictionariesGallery(){
        Intent intent = new Intent(SearchActivity.this, DictionaryActivity.class);
        startActivity(intent);
    }

    @OnItemSelected(R.id.spResults)
    void onItemSelected(int pos) {
        Log.d(getClass().getName(), "item selected " + pos);
        SearchResult result = (SearchResult) spResults.getItemAtPosition(pos);
        selectResult(result);
    }

    @OnClick(R.id.ivSwapLanguages)
    public void swapLanguages(View v){
        this.isFromQuechua = !this.isFromQuechua;
        this.updateViewsOnLangChange();
        preferencesHelper.saveSearchFromQuechua(this.isFromQuechua);
    }

    @OnItemSelected(R.id.spDictLangs)
    void onDictLangSelected() {
        updateQueryHint();
        preferencesHelper.saveNonQuechuaLangPos(((DictLang)spDictLangs.getSelectedItem()).getCode());
    }

    private void setSearchParams(SearchParams searchParams) {
        this.searchParams = searchParams;
        this.isFromQuechua = searchParams.isFromQuechua();
        if(!this.isFromQuechua) {
            this.updateViewsOnLangChange();
        }
        spSearchTypes.setSelection(searchParams.getSearchTypePos());
        int dictLangPos = ((DictLangAdapter)spDictLangs.getAdapter()).getPosByValue(searchParams.getNonQuechuaLangCode());
        spDictLangs.setSelection(Math.max(0, dictLangPos));
        updateQueryHint();
    }

    private void updateViewsOnLangChange() {
        llSearchOptions.removeAllViews();
        List<View> innerViews = Arrays.asList(tvQichwaLang, ivSwapLanguages, spDictLangs);
        if(!isFromQuechua) {
            Collections.reverse(innerViews);
        }
        llSearchOptions.addView(spSearchTypes);
        llSearchOptions.addView(vSeparator);
        for(View view : innerViews) {
            llSearchOptions.addView(view);
        }
        updateQueryHint();
    }

    private void updateQueryHint() {
        if(isFromQuechua) {
            svSearch.setQueryHint(quechuaPlaceholder);
        } else {
            DictLang dictLang = (DictLang) spDictLangs.getSelectedItem();
            svSearch.setQueryHint(placeholders.get(dictLang.getCode()));
        }
    }

    private void onResultsChanged(List<SearchResult> searchResults){
        llResultsArea.setVisibility(View.VISIBLE);
        if(searchResults == null  || searchResults.isEmpty()){
            llNoResults.setVisibility(View.VISIBLE);
            spResults.setVisibility(View.GONE);
            tvResultsTotal.setVisibility(View.GONE);
            rvResults.setVisibility(View.GONE);
            spinnerResultAdapter.setSearchResults(new ArrayList<>());
            resultListAdapter.setDefinitions(new ArrayList<>());
        }else{
            llNoResults.setVisibility(View.GONE);
            spResults.setVisibility(View.VISIBLE);
            tvResultsTotal.setVisibility(View.VISIBLE);
            renderResults(searchResults);
        }
    }

    private void onSearchViewStateChanged(SearchViewState searchViewState){
        if (searchViewState != null) {
            if(searchViewState.isLoading()){
                searchProgressBar.setVisibility(View.VISIBLE);
                llResultsArea.setVisibility(View.GONE);
            }else{
                searchProgressBar.setVisibility(View.GONE);
            }
        }
    }

    private void onFetchExtraDefinitions(List<Definition> definitions){
        if(definitions != null && !definitions.isEmpty()){
            if(resultListAdapter != null){
                int currentLength = resultListAdapter.getItemCount();
                resultListAdapter.pushMoreDefinitions(definitions);
                resultListAdapter.notifyItemRangeInserted(currentLength, definitions.size());
            }
        }
    }

    private void onFetchMoreLoadingChanged(Boolean isLoading){
        fetchMoreProgressBar.setVisibility(isLoading == null || !isLoading ? View.GONE : View.VISIBLE);
    }

    private void onSaveWordResult(Boolean isSuccess){
        int stringId = (isSuccess != null && isSuccess) ? R.string.favorite_added_success : R.string.favorite_added_error;
        Snackbar.make(clSearch, getString(stringId), Snackbar.LENGTH_SHORT).show();
    }

    private void renderResults(List<SearchResult> searchResults) {
        searchType = spSearchTypes.getSelectedItemPosition();
        int previousSelection = spResults.getSelectedItemPosition();

        spinnerResultAdapter.setSearchResults(searchResults);
        spinnerResultAdapter.notifyDataSetChanged();
        spResults.setSelection(0, true);
        if(previousSelection == 0){
            selectResult(Objects.requireNonNull(spinnerResultAdapter.getItem(0)));
        }
        Log.d(getClass().getName(), "set results, position = " + previousSelection);
    }

    private void selectResult(SearchResult result){
        rvResults.scrollToPosition(0);
        scrollListener.resetState();
        rvResults.setVisibility(View.VISIBLE);
        tvResultsTotal.setText(getResources().getQuantityString(R.plurals.total_results, result.getTotal(),
                result.getTotal()));
        resultListAdapter.setDefinitions(result.getDefinitions());
        resultListAdapter.notifyDataSetChanged();
    }

    private void loadPlaceholders(){
        this.placeholders = new HashMap<>();
        this.quechuaPlaceholder = getString(R.string.search_quechua_placeholder);
        this.placeholders.put("qu", quechuaPlaceholder);
        this.placeholders.put("es", getString(R.string.search_spanish_placeholder));
        this.placeholders.put("en", getString(R.string.search_english_placeholder));
        this.placeholders.put("fr", getString(R.string.search_french_placeholder));
        this.placeholders.put("de", getString(R.string.search_german_placeholder));
        this.placeholders.put("it", getString(R.string.search_italian_placeholder));
    }

    private void setAdapters(){
        ArrayAdapter<CharSequence> searchTypesAdapter = ArrayAdapter.createFromResource(this, R.array.searchTypes, R.layout.item_spinner_search_type);
        searchTypesAdapter.setDropDownViewResource(R.layout.item_spinner_dropdown_search_type);
        spSearchTypes.setAdapter(searchTypesAdapter);
        spSearchTypes.setSelection(2);

        spDictLangs.setAdapter(new DictLangAdapter(this));
        spinnerResultAdapter = new SpinnerResultAdapter(this, R.layout.item_spinner_result, new ArrayList<>());
        resultListAdapter = new ResultListAdapter(new ArrayList<>(), this);
        rvResults.setAdapter(resultListAdapter);
        spResults.setAdapter(spinnerResultAdapter);
    }

    private String getCurrentLang(){
        return ((DictLang)spDictLangs.getSelectedItem()).getCode();
    }



    private void setSearchView(){
        svSearch.setQuery(svSearch.getQueryHint(), false);
        svSearch.setOnQueryTextListener(new SearchView.OnQueryTextListener() {
            @Override
            public boolean onQueryTextSubmit(String query) {
                Log.d(getClass().getName(), "onquerysubmit " + query);
                viewModel.searchWord(isFromQuechua ? 1 : 0, getCurrentLang(), spSearchTypes.getSelectedItemPosition(), query);
                svSearch.clearFocus();
                return true;
            }

            @Override
            public boolean onQueryTextChange(String newText) {
                return true;
            }
        });
    }


}

File: app/src/main/java/com/ocram/qichwadic/features/search/ui/SearchActivity.kt
package com.ocram.qichwadic.features.search.ui

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.SearchView
import androidx.appcompat.widget.Toolbar
import androidx.core.text.HtmlCompat
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

import com.google.android.material.snackbar.Snackbar
import com.ocram.qichwadic.R
import com.ocram.qichwadic.features.common.data.model.DefinitionEntity
import com.ocram.qichwadic.features.common.domain.SearchResultModel
import com.ocram.qichwadic.features.about.ui.AboutActivity
import com.ocram.qichwadic.features.dictionaries.ui.DictionaryActivity
import com.ocram.qichwadic.features.favorites.ui.FavoriteActivity
import com.ocram.qichwadic.features.search.ui.custom.EndlessRecyclerViewScrollListener
import com.ocram.qichwadic.features.common.DictLang
import com.ocram.qichwadic.features.common.SearchParams
import com.ocram.qichwadic.core.ui.activity.BaseActivity
import com.ocram.qichwadic.features.common.domain.DefinitionModel

import java.util.ArrayList

import org.koin.androidx.viewmodel.ext.android.viewModel
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.toolbar.*
import kotlin.math.max

class SearchActivity : BaseActivity(), SearchResultListAdapter.DefinitionClickListener {

    private lateinit var spinnerResultAdapter: SpinnerResultAdapter
    private lateinit var searchResultListAdapter: SearchResultListAdapter
    private lateinit var scrollListener: EndlessRecyclerViewScrollListener
    private var searchType: Int = 0
    private lateinit var placeholders: Map<String, String>
    private lateinit var quechuaPlaceholder: String
    private var isFromQuechua: Boolean = false
    private var searchParams: SearchParams? = null

    private val searchViewModel: SearchViewModel by viewModel()
    private var offlineSearch: Boolean = false

    override val layoutId: Int = R.layout.activity_main

    override fun getToolbar(): Toolbar? {
        return mToolbar as Toolbar?
    }

    private lateinit var currentLang: String

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        loadPlaceholders()
        setAdapters()
        initSearchView()
        currentLang = (spDictLangs.selectedItem as DictLang).code
        searchViewModel.resultLiveData.observe(this, Observer<List<SearchResultModel>> { this.onResultsChanged(it) })
        searchViewModel.searchViewState.observe(this, Observer<SearchViewState> { this.onSearchViewStateChanged(it) })
        searchViewModel.extraDefinitions.observe(this, Observer<List<DefinitionModel>> { this.onFetchExtraDefinitions(it) })
        searchViewModel.loadFetchMore.observe(this, Observer<Boolean> { this.onFetchMoreLoadingChanged(it) })
        searchViewModel.saveFavoriteResult.observe(this, Observer<Boolean> { this.onSaveWordResult(it) })
        searchViewModel.offlineSearch.observe(this, Observer<Boolean> { this.onSearchModeChanged(it) })
        searchViewModel.searchParams.observe(this, Observer<SearchParams> { this.setSearchParams(it) })
    }

    private fun loadPlaceholders() {
        this.quechuaPlaceholder = getString(R.string.search_quechua_placeholder)
        this.placeholders = mapOf(
                Pair("qu", quechuaPlaceholder),
                Pair("es", getString(R.string.search_spanish_placeholder)),
                Pair("en", getString(R.string.search_english_placeholder)),
                Pair("fr", getString(R.string.search_french_placeholder)),
                Pair("de", getString(R.string.search_german_placeholder)),
                Pair("it", getString(R.string.search_italian_placeholder)),
                Pair("ru", getString(R.string.search_russian_placeholder))
        )
    }

    private fun setAdapters() {
        val searchTypesAdapter = ArrayAdapter.createFromResource(this, R.array.searchTypes, R.layout.item_spinner_search_type)
        searchTypesAdapter.setDropDownViewResource(R.layout.item_spinner_dropdown_search_type)
        spSearchTypes.adapter = searchTypesAdapter
        spSearchTypes.setSelection(2)

        spDictLangs.adapter = SearchDictLangAdapter(this)
        spinnerResultAdapter = SpinnerResultAdapter(this, R.layout.item_spinner_result, ArrayList())
        searchResultListAdapter = SearchResultListAdapter(ArrayList(), this)
        rvResults.adapter = searchResultListAdapter
        spResults.adapter = spinnerResultAdapter
    }

    private fun initSearchView() {
        svSearch.setQuery(svSearch.queryHint, false)
        svSearch.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextSubmit(query: String): Boolean {
                Log.d(javaClass.name, "onquerysubmit $query")
                searchViewModel.searchWord(if (isFromQuechua) 1 else 0, (spDictLangs.selectedItem as DictLang).code, spSearchTypes.selectedItemPosition, query)
                svSearch.clearFocus()
                return true
            }

            override fun onQueryTextChange(newText: String): Boolean {
                return true
            }
        })
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_toolbar, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {

        when (item.itemId) {
            R.id.nav_dictionaries -> {
                openDictionariesGallery()
                return true
            }
            R.id.nav_favorites -> {
                val intent = Intent(this@SearchActivity, FavoriteActivity::class.java)
                startActivity(intent)
                return true
            }
            R.id.nav_about -> {
                val intent = Intent(this@SearchActivity, AboutActivity::class.java)
                startActivity(intent)
                return true
            }
            else -> return super.onOptionsItemSelected(item)
        }
    }

    override fun initViews() {
        btnNoResultsGetDictionaries.setOnClickListener { openDictionariesGallery() }
        ivSwapLanguages.setOnClickListener { swapLanguages() }
        fabSearch.setOnClickListener { runSearch() }
        spResults.onItemSelectedListener = object: AdapterView.OnItemSelectedListener {
            override fun onNothingSelected(p0: AdapterView<*>?) {}

            override fun onItemSelected(p0: AdapterView<*>?, p1: View?, pos: Int, p3: Long) {
                onItemSelected(pos)
            }
        }

        spDictLangs.onItemSelectedListener = object: AdapterView.OnItemSelectedListener {
            override fun onNothingSelected(p0: AdapterView<*>?) {}

            override fun onItemSelected(p0: AdapterView<*>?, p1: View?, pos: Int, p3: Long) {
                onDictLangSelected()
            }

        }

        swOfflineSearch.setOnCheckedChangeListener { _, checked -> onSearchOfflineModeChanged(checked) }
        ivSubmitSearch.setOnClickListener { runSearch() }
        setToolbar()
        setRecyclerViews()
    }

    private fun openDictionariesGallery() {
        val intent = Intent(this@SearchActivity, DictionaryActivity::class.java)
        startActivity(intent)
    }

    private fun runSearch() {
        if (svSearch.query.isNotEmpty()) {
            svSearch.setQuery(svSearch.query, true)
        }
    }

    private fun swapLanguages() {
        this.isFromQuechua = !this.isFromQuechua
        this.updateViewsOnLangChange()
        searchViewModel.saveSearchFromQuechua(this.isFromQuechua)
    }

    private fun onItemSelected(pos: Int) {
        Log.d(javaClass.name, "item selected $pos")
        val result = spResults.getItemAtPosition(pos) as SearchResultModel
        selectResult(result)
    }

    private fun onDictLangSelected() {
        updateQueryHint()
        spDictLangs.selectedItem.let {
            searchViewModel.saveNonQuechuaLangPos((it as DictLang).code)
        }
    }

    private fun onSearchOfflineModeChanged(checked: Boolean) {
        searchViewModel.changeSearchModeConfig(checked)
    }

    private fun setToolbar() {
        if (supportActionBar != null) {
            supportActionBar!!.setDisplayShowTitleEnabled(true)
            supportActionBar!!.setHomeButtonEnabled(false)
            supportActionBar!!.setDisplayHomeAsUpEnabled(false)
        }
    }

    private fun setRecyclerViews() {
        val layoutManager = LinearLayoutManager(this, RecyclerView.VERTICAL, false)
        rvResults.layoutManager = layoutManager
        scrollListener = object : EndlessRecyclerViewScrollListener(layoutManager) {
            override fun onLoadMore(page: Int, totalItemsCount: Int, view: RecyclerView) {
                val currentSearchResult = spResults.selectedItem as SearchResultModel
                val totalToFetch = currentSearchResult.total
                if (totalToFetch > 20 && searchResultListAdapter.itemCount < totalToFetch) {
                    val dictionaryId = currentSearchResult.dictionaryId
                    searchViewModel.fetchMoreResults(dictionaryId, searchType, svSearch.query.toString(), page + 1)
                }
            }
        }
        rvResults.addOnScrollListener(scrollListener)
    }

    override fun onItemShareClick(definition: DefinitionModel) {
        val text = getString(R.string.share_definition_from_dictionary,
                definition.dictionaryName, definition.word, definition.meaning)
        openShareIntent(HtmlCompat.fromHtml(text, HtmlCompat.FROM_HTML_MODE_LEGACY).toString())
    }

    override fun onItemFavoriteClick(definition: DefinitionModel) {
        searchViewModel.saveFavorite(definition)
    }

    private fun setSearchParams(searchParams: SearchParams) {
        this.searchParams = searchParams
        this.isFromQuechua = searchParams.isFromQuechua
        if (!this.isFromQuechua) {
            this.updateViewsOnLangChange()
        }
        spSearchTypes.setSelection(searchParams.searchTypePos)
        val dictLangPos = searchParams.nonQuechuaLangCode?.let { (spDictLangs.adapter as SearchDictLangAdapter).getPosByValue(it) }
        dictLangPos?.let { max(0, it) }?.let { spDictLangs.setSelection(it) }
        updateQueryHint()
        svSearch.setQuery(svSearch.queryHint, true)
    }

    private fun updateViewsOnLangChange() {
        val innerViews = mutableListOf(spSearchTypes, vSeparator, tvQichwaLang, ivSwapLanguages, spDictLangs)
        if(!isFromQuechua) {
            innerViews.removeAt(4)
            innerViews.removeAt(2)
            innerViews.add(2, spDictLangs)
            innerViews.add(4, tvQichwaLang)
        }
        llSearchOptions.removeAllViews()
        innerViews.forEach { llSearchOptions.addView(it) }
        updateQueryHint()
    }

    private fun updateQueryHint() {
        if (isFromQuechua) {
            svSearch.queryHint = quechuaPlaceholder
        } else {
            this.currentLang = (spDictLangs.selectedItem as DictLang).code
            svSearch.queryHint = placeholders[this.currentLang]
        }
    }

    private fun onResultsChanged(searchResults: List<SearchResultModel>?) {
        llResultsArea.visibility = View.VISIBLE
        tvError.visibility = View.GONE
        if (searchResults == null || searchResults.isEmpty()) {
            spinnerResultAdapter.setSearchResults(emptyList())
            searchResultListAdapter.setDefinitions(mutableListOf())
            spResults.visibility = View.GONE
            tvResultsTotal.visibility = View.GONE
            rvResults.visibility = View.GONE
            if (offlineSearch) {
                llNoResultsForOffline.visibility = View.VISIBLE
            } else {
                llNoResultsForOnline.visibility = View.VISIBLE
            }
        } else {
            llNoResultsForOffline.visibility = View.GONE
            llNoResultsForOnline.visibility = View.GONE
            spResults.visibility = View.VISIBLE
            tvResultsTotal.visibility = View.VISIBLE
            renderResults(searchResults)
        }
    }

    private fun onSearchViewStateChanged(searchViewState: SearchViewState) {
        when(searchViewState) {
            SearchViewState.LOADING -> {
                searchProgressBar.visibility = View.VISIBLE
                llResultsArea.visibility = View.GONE
                llNoResultsForOffline.visibility = View.GONE
                llNoResultsForOnline.visibility = View.GONE
                tvError.visibility = View.GONE
            }
            SearchViewState.SUCCESS -> {
                searchProgressBar.visibility = View.GONE
                tvError.visibility = View.GONE
            }
            SearchViewState.ERROR -> {
                searchProgressBar.visibility = View.GONE
                val errorMsgId = if (offlineSearch) R.string.error_offline else R.string.error_online
                tvError.text = getString(errorMsgId)

                llResultsArea.visibility = View.GONE
                llNoResultsForOffline.visibility = View.GONE
                llNoResultsForOnline.visibility = View.GONE
                tvError.visibility = View.VISIBLE
            }
        }
    }

    private fun onFetchExtraDefinitions(definitions: List<DefinitionModel>?) {
        if (!definitions.isNullOrEmpty()) {
            val currentLength = searchResultListAdapter.itemCount
            searchResultListAdapter.pushMoreDefinitions(definitions)
            searchResultListAdapter.notifyItemRangeInserted(currentLength, definitions.size)
        }
    }

    private fun onFetchMoreLoadingChanged(isLoading: Boolean?) {
        fetchMoreProgressBar.visibility = if (isLoading == null || !isLoading) View.GONE else View.VISIBLE
    }

    private fun onSaveWordResult(isSuccess: Boolean?) {
        val stringId = if (isSuccess != null && isSuccess) R.string.favorite_added_success else R.string.favorite_added_error
        Snackbar.make(clSearch, getString(stringId), Snackbar.LENGTH_SHORT).show()
    }

    private fun onSearchModeChanged(isOffline: Boolean?) {
        this.offlineSearch = isOffline != null && isOffline
        swOfflineSearch.isChecked = this.offlineSearch
        this.invalidateOptionsMenu()
    }

    private fun renderResults(searchResults: List<SearchResultModel>) {
        searchType = spSearchTypes.selectedItemPosition
        val previousSelection = spResults.selectedItemPosition

        spinnerResultAdapter.setSearchResults(searchResults)
        spinnerResultAdapter.notifyDataSetChanged()
        spResults.setSelection(0, true)
        if (previousSelection == 0) {
            spinnerResultAdapter.getItem(0)?.let { selectedItem -> selectResult(selectedItem) }
        }
    }

    private fun selectResult(result: SearchResultModel) {
        rvResults.scrollToPosition(0)
        scrollListener.resetState()
        rvResults.visibility = View.VISIBLE
        tvResultsTotal.text = resources.getQuantityString(R.plurals.total_results, result.total,
                result.total)
        searchResultListAdapter.setDefinitions(result.definitions)
        searchResultListAdapter.notifyDataSetChanged()
    }

}
